### PR TITLE
8264973: AArch64: Optimize vector max/min/add reduction of two integers with NEON pairwise instructions

### PR DESCRIPTION
**Pull request by dgbo**
On aarch64, current implementations of vector reduce_add2I, reduce_max2I, reduce_min2I can be optimized with NEON pairwise instructions:

```
## reduce_add2I, before
mov    w10, v19.s[0]
mov    w2, v19.s[1]
add    w10, w0, w10
add    w10, w10, w2
## reduce_add2I, optimized
addp   v23.2s, v24.2s, v24.2s
mov    w10, v23.s[0]
add    w10, w10, w2

## reduce_max2I, before
dup    v16.2d, v23.d[0]
sminv  s16, v16.4s
mov    w10, v16.s[0]
cmp    w10, w0
csel   w10, w10, w0, lt
## reduce_max2I, optimized
sminp  v16.2s, v23.2s, v23.2s
mov    w10, v16.s[0]
cmp    w10, w0
csel   w10, w10, w0, lt
```

I don't expect this to change anything of SuperWord, vectorizing reductions of two integers is disabled by [1].
This is useful for VectorAPI, tested benchmarks in [2], performance can improve ~51% and ~8% for `Int64Vector.ADD` and `Int64Vector.MAX` respectively.

```
Benchmark                   (size)   Mode  Cnt     Score    Error   Units
# optimized
Int64Vector.ADDLanes          1024  thrpt   10  2492.123 ± 23.561  ops/ms
Int64Vector.ADDMaskedLanes    1024  thrpt   10  1825.882 ±  5.261  ops/ms
Int64Vector.MAXLanes          1024  thrpt   10  1921.028 ±  3.253  ops/ms
Int64Vector.MAXMaskedLanes    1024  thrpt   10  1588.575 ±  3.903  ops/ms
Int64Vector.MINLanes          1024  thrpt   10  1923.913 ±  2.117  ops/ms
Int64Vector.MINMaskedLanes    1024  thrpt   10  1596.875 ±  2.163  ops/ms
# default
Int64Vector.ADDLanes          1024  thrpt   10  1644.223 ±  1.885  ops/ms
Int64Vector.ADDMaskedLanes    1024  thrpt   10  1491.502 ± 26.436  ops/ms
Int64Vector.MAXLanes          1024  thrpt   10  1784.066 ±  3.816  ops/ms
Int64Vector.MAXMaskedLanes    1024  thrpt   10  1494.750 ±  3.451  ops/ms
Int64Vector.MINLanes          1024  thrpt   10  1785.266 ±  8.893  ops/ms
Int64Vector.MINMaskedLanes    1024  thrpt   10  1499.233 ±  3.498  ops/ms
```

Verified correctness with tests `test/jdk/jdk/incubator/vector/`. Also tested linux-aarch64-server-fastdebug tier1-3.

[1] https://github.com/openjdk/jdk/blob/3bf4c904fbbd87d4db18db22c1be384616483eed/src/hotspot/share/opto/superword.cpp#L2004
[2] https://github.com/openjdk/panama-vector/blob/vectorIntrinsics/test/jdk/jdk/incubator/vector/benchmark/src/main/java/benchmark/jdk/incubator/vector/Int64Vector.java

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264973](https://bugs.openjdk.java.net/browse/JDK-8264973): AArch64: Optimize vector max/min/add reduction of two integers with NEON pairwise instructions

### Reviewers
 * [Ningsheng Jian](https://openjdk.java.net/census#njian) (@nsjian - Committer) ⚠️ Review applies to 0de0bef46d24087636284f245e3e94cd458f5b2b
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3683/head:pull/3683` \
`$ git checkout pull/3683`

Update a local copy of the PR: \
`$ git checkout pull/3683` \
`$ git pull https://git.openjdk.java.net/jdk pull/3683/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3683`

View PR using the GUI difftool: \
`$ git pr show -t 3683`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3683.diff">https://git.openjdk.java.net/jdk/pull/3683.diff</a>

</details>

### CODE REVIEW

**Review by nsjian — APPROVED** [review]
Yes, this looks good to me. Thanks!

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 1]
```diff
@@ -2404,6 +2404,8 @@ void mvnw(Register Rd, Register Rm,
   INSN(umlalv, 1, 0b100000, false); // accepted arrangements: T8B, T16B, T4H, T8H, T2S, T4S
   INSN(maxv,   0, 0b011001, false); // accepted arrangements: T8B, T16B, T4H, T8H, T2S, T4S
   INSN(minv,   0, 0b011011, false); // accepted arrangements: T8B, T16B, T4H, T8H, T2S, T4S
+  INSN(smaxp,  0, 0b101001, false); // accepted arrangements: T8B, T16B, T4H, T8H, T2S, T4S
+  INSN(sminp,  0, 0b101011, false); // accepted arrangements: T8B, T16B, T4H, T8H, T2S, T4S
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:2408]: Just one nit: do we need an assembler test case for them?

Comment (dgbo) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:2408]: Sure, added in the updated commit. Thanks.

**Comment by dgbo** [comment]
Hi, @theRealAph. Could you please take a look at this PR? Thanks.

**Comment by dgbo** [comment]
PING... Any comments/suggestions are appreciated.
Although this has been reviewed by Ningsheng, we still need help from reviewers here.

**Comment by PaulSandoz** [comment]
@theRealAph we are still working out how best to bring the vector performance tests over to the `test/micro` area of mainline. (Some preliminary work is [here](https://github.com/openjdk/panama-vector/pull/77)). The perf tests in the `panama-vector` are under a maven project and it should be possible to build/run that project with a mainline build of the JDK (the tests should be compatible).

**Comment by dgbo** [comment]
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-compiler-dev](mailto:hotspot-compiler-dev@mail.openjdk.java.net):_
> 
> On 5/10/21 6:55 AM, Dong Bo wrote:
> 
> > PING? Any comments/suggestions are appreciated.
> > Although this has been reviewed by Ningsheng, we still need help from reviewers here.
> 
> I'm testing this now.

PING? Any suggestions? Thanks.

**Comment by theRealAph** [comment]
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-compiler-dev](mailto:hotspot-compiler-dev@mail.openjdk.java.net):_
> 
> On 5/10/21 6:55 AM, Dong Bo wrote:
> 
> > PING? Any comments/suggestions are appreciated.
> > Although this has been reviewed by Ningsheng, we still need help from reviewers here.
> 
> I'm testing this now.

I'm back on this, and I can't see how to run the tests. Paul Sandoz says the perf tests in the `panama-vector` are under a maven project, but which maven project is that? 

You say "When I tested this, the incompatible DecodeBench.java was deleted first since it is all about ShortVector and ByteVector rather than Int64Vector." but I don't know what that means.

Please provide step-by-step instructions that allow anyone to reproduce your results.

**Comment by dgbo** [comment]
> > _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-compiler-dev](mailto:hotspot-compiler-dev@mail.openjdk.java.net):_
> > On 5/10/21 6:55 AM, Dong Bo wrote:
> > > PING? Any comments/suggestions are appreciated.
> > > Although this has been reviewed by Ningsheng, we still need help from reviewers here.
> > 
> > 
> > I'm testing this now.
> 
> I'm back on this, and I can't see how to run the tests. Paul Sandoz says the perf tests in the `panama-vector` are under a maven project, but which maven project is that?
> 
> You say "When I tested this, the incompatible DecodeBench.java was deleted first since it is all about ShortVector and ByteVector rather than Int64Vector." but I don't know what that means.
> 
> Please provide step-by-step instructions that allow anyone to reproduce your results.

Hi, here are instructions I used to run the benchmarks:
1. Get the benchmark project in `https://github.com/openjdk/panama-vector/tree/vectorIntrinsics/test/jdk/jdk/incubator/vector/benchmark` (I do this via git):
```
## git clone https://github.com/openjdk/panama-vector.git
## cd panama-vector
## git checkout -b vectorIntrinsics remotes/origin/vectorIntrinsics
```
2. Delete incompatible `DecodeBench.java` and compile the benchmarks with mainline JDK:
```
## <set JAVA_HOME to the mainline JDK path>
## cd test/jdk/jdk/incubator/vector/benchmark
## rm src/main/java/benchmark/utf8/DecodeBench.java
## mvn install
```
3. Run tests:
`## <PATH_TO_TEST_JDK>/bin/java -jar target/vector-benchmarks.jar benchmark.jdk.incubator.vector.Int64Vector.["M"\|"ADD"]+[AXIN]*["Masked"]*Lanes -wi 10 -w 1000ms -f 1 -i 10 -r 1000ms`

**Review by theRealAph — APPROVED** [review]
Thanks. Changes look good.

I managed to reproduce your results.
```

Benchmark                   (size)   Mode  Cnt     Score    Error   Units
Int64Vector.ADDLanes          1024  thrpt    3  4958.747 ± 54.225  ops/ms
Int64Vector.ADDMaskedLanes    1024  thrpt    3  4769.759 ± 12.736  ops/ms
Int64Vector.MAXLanes          1024  thrpt    3  2957.985 ± 88.671  ops/ms
Int64Vector.MAXMaskedLanes    1024  thrpt    3  2921.381 ± 45.408  ops/ms
Int64Vector.MINLanes          1024  thrpt    3  2965.392 ± 25.236  ops/ms
Int64Vector.MINMaskedLanes    1024  thrpt    3  2923.870 ± 53.270  ops/ms

Benchmark                   (size)   Mode  Cnt     Score    Error   Units
Int64Vector.ADDLanes          1024  thrpt    3  3560.100 ± 79.753  ops/ms
Int64Vector.ADDMaskedLanes    1024  thrpt    3  3585.672 ± 57.203  ops/ms
Int64Vector.MAXLanes          1024  thrpt    3  2951.659 ±  9.577  ops/ms
Int64Vector.MAXMaskedLanes    1024  thrpt    3  2876.957 ± 37.005  ops/ms
Int64Vector.MINLanes          1024  thrpt    3  2953.476 ±  3.446  ops/ms
Int64Vector.MINMaskedLanes    1024  thrpt    3  2878.942 ± 50.281  ops/ms

```

**Pull request closed without merging**